File: src/lib/backend-client/stars.ts
/**
 * Typed front-end client for the stars (favorites / 加星) commands.
 *
 * Why this file exists:
 * - PathKeep stores stars against canonical entities in the archive
 *   (migration 014 + `vault-core::stars`). Explorer rows, the detail panel,
 *   search results, and assistant evidence toggle them; the Starred hub reads
 *   them.
 * - Keeping the typed client here means routes never type raw command names
 *   ("set_star") and stay shielded against transport renames.
 *
 * Main declarations:
 * - `starsClient`
 *
 * Source-of-truth notes:
 * - Transport contract: `docs/architecture/desktop-command-surface.md`.
 * - Shape contract: `vault_core::models::stars` (StarEntityKind, StarSort,
 *   SetStarRequest, StarStatusRequest, StarListItem, StarCounts).
 */

import { call } from './shared'

/** The kind of entity a star points at. `queryFamily` is deferred. */
export type StarEntityKind = 'url' | 'domain'

/** Ordering for the Starred hub. */
export type StarSort = 'recently_starred' | 'most_revisited'

/** A request to add or remove a star. */
export interface SetStarRequest {
  entityKind: StarEntityKind
  entityKey: string
  sourceProfile?: string | null
}

/** A batched status request for the currently-visible rows only. */
export interface StarStatusRequest {
  entityKind: StarEntityKind
  entityKeys: string[]
}

/** One starred entity returned by `list_stars`, enriched for the hub. */
export interface StarListItem {
  entityKind: StarEntityKind
  entityKey: string
  starredAt: string
  domain: string
  title: string
  visitCount: number
}

/** Per-kind rollup of how many things the user has starred. */
export interface StarCounts {
  urls: number
  domains: number
}

export const starsClient = {
  setStar: (request: SetStarRequest) => call<void>('set_star', { request }),
  unsetStar: (request: SetStarRequest) => call<void>('unset_star', { request }),
  /** Map of the supplied raw keys → starred boolean. Visible rows only. */
  getStarStatus: (request: StarStatusRequest) =>
    call<Record<string, boolean>>('get_star_status', { request }),
  listStars: (kind: StarEntityKind | null, sort: StarSort, limit?: number) =>
    call<StarListItem[]>('list_stars', {
      kind: kind ?? null,
      sort,
      limit: limit ?? null,
    }),
  getStarCounts: () => call<StarCounts>('get_star_counts', {}),
}

File: src-tauri/crates/vault-core/src/models/stars.rs
//! Star (favorites / 加星) models.
//!
//! A star marks a canonical entity — a page (`canonical_url`) or a source
//! (`registrable_domain`) — as a user favorite. These serde types are the
//! contract between the desktop shell, the worker layer, and the `stars`
//! backend module (migration 014). They mirror the annotations models: keyed
//! by the canonical entity, profile-agnostic, and portable across the export
//! bundle.

use serde::{Deserialize, Serialize};

/// The kind of entity a star points at. The MVP supports pages (`url`) and
/// sources (`domain`); `query_family` stars are deferred so the enum can grow
/// without breaking the on-disk `entity_kind` text values.
#[derive(Clone, Copy, Debug, Default, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "snake_case")]
pub enum StarEntityKind {
    /// A single page, keyed by its `canonical_url`.
    #[default]
    Url,
    /// A whole source, keyed by its `registrable_domain`.
    Domain,
}

impl StarEntityKind {
    /// Returns the stable text value persisted in the `star.entity_kind`
    /// column. This is the single source of truth for the SQLite encoding —
    /// keep it aligned with the `#[serde]` rename so the wire and storage
    /// forms never drift.
    pub fn as_str(self) -> &'static str {
        match self {
            StarEntityKind::Url => "url",
            StarEntityKind::Domain => "domain",
        }
    }
}

/// How `list_stars` orders the Starred hub.
#[derive(Clone, Copy, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum StarSort {
    /// Newest stars first (`starred_at DESC`) — rides the kind index.
    #[default]
    RecentlyStarred,
    /// Most-revisited first, via a join to the per-URL visit-count read model.
    MostRevisited,
}

/// A request to add or remove a star. `entity_key` is the `canonical_url`
/// (for `Url`) or the `registrable_domain` (for `Domain`); the backend
/// canonicalizes raw URLs before keying so callers may pass either.
///
/// Stars key by canonical_url BY DESIGN (annotations key by RAW url): a star is
/// page identity, so tracking-param + host-casing variants collapse onto one
/// canonical key and the star survives re-import. `set_star`, `unset_star`, and
/// `is_starred_batch` all canonicalize the supplied key before reading/writing.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SetStarRequest {
    pub entity_kind: StarEntityKind,
    pub entity_key: String,
    /// Audit-only profile id captured at write time. Never part of the key.
    #[serde(default)]
    pub source_profile: Option<String>,
}

/// A request to read star status for the currently-visible rows only. The
/// frontend batches one call per render window so the lookup never fans out
/// across the whole archive.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "camelCase")]
pub struct StarStatusRequest {
    pub entity_kind: StarEntityKind,
    pub entity_keys: Vec<String>,
}

/// One starred entity returned by `list_stars`, enriched with the display
/// fields the Starred hub renders without a second round-trip.
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct StarListItem {
    pub entity_kind: StarEntityKind,
    pub entity_key: String,
    /// RFC 3339 timestamp the star was created.
    pub starred_at: String,
    /// Registrable domain for the entity — equals `entity_key` for domain
    /// stars, derived from the canonical URL for page stars. Empty when it
    /// cannot be derived.
    #[serde(default)]
    pub domain: String,
    /// Best-known page title (page stars only), resolved from the canonical
    /// page (i.e. across every raw-URL variant that collapses to the star key).
    /// Empty when the archive has no row for the page yet.
    #[serde(default)]
    pub title: String,
    /// Total visit count for the entity, SUMMED across every raw-URL variant of
    /// the canonical page (or every URL on the domain). Drives the
    /// most-revisited sort and the hub's "{n}×" affordance.
    #[serde(default)]
    pub visit_count: i64,
}

/// Rollup of how many things the user has starred, per kind.
#[derive(Clone, Copy, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct StarCounts {
    pub urls: i64,
    pub domains: i64,
}
